pkg/domain/storage.go

```diff
@@ -2,9 +2,11 @@ package domain
 
 import (
 	"context"
+	"time"
 )
 
 // Storage interface provides functions for inserting and retrieving assets
 type Storage interface {
 	StoreCloudAsset(context.Context, CloudAssetChanges) error
+	FetchCloudAsset(ctx context.Context, hostname string, ipAddress string, timestamp time.Time) (CloudAssetDetails, error)
```

Comment:
Similar to my API comments. I think we should start with highly targeted interfaces for attributes rather than the "catch-all". As the number of searchable values for an asset grows, this interface will become unwieldy. At least in the beginning, I'd rather see things like:
```golang
type IPFetcherBy interface {
   FetchByIP(ctx context.Context, when time.Time, ipAddress string) (CloudAssetDetails, error)
}
```
This will both enable us to focus only on implementing the queries from a single perspective (rather than exposing a dynamic query function for the entire database) as well as enable new features to be developed without modifying one, monolithic feature function. Again, I'm OK drifting towards a more complex/powerful system later but let's keep it constrained for now.

Reply:
Does each query need its own interface, or would it be sufficient to separate queries on each attribute to separate functions within a single interface? For example, splitting the `Storage` interface into something like:

```go
type CloudAssetStorer interface {
    Store(context.Context, CloudAssetChanges) error
}

type CloudAssetFetcher interface {
    FetchByIP(ctx context.Context, when time.Time, ipAddress string) (CloudAssetDetails, error)
    FetchByHostname(ctx context.Context, when time.Time, hostname string) (CloudAssetDetails, error)
}
```